libtenzir/builtins/operators/load_tcp.cpp

```diff
@@ -0,0 +1,864 @@
+    if (buffer.size() < max_buffered_batches) {
+      return {};
+    }
+    return write_rps.emplace(self->template make_response_promise<void>());
+  }
+
+  auto read_elements() -> caf::result<Elements> {
+    TENZIR_ASSERT(not read_rp.pending());
+    if (not buffer.empty()) {
+      auto elements = std::move(buffer.front());
+      buffer.pop();
+      while (buffer.size() < max_buffered_batches and not write_rps.empty()) {
+        auto write_rp = std::move(write_rps.front());
+        write_rps.pop();
+        TENZIR_ASSERT(write_rp.pending());
+        write_rp.deliver();
+      }
+      return std::move(elements);
+    }
+    read_rp = self->template make_response_promise<Elements>();
+    return read_rp;
+  }
+
+  auto handle_down_msg(const caf::down_msg& msg) -> void {
+    const auto connection
+      = std::ranges::find_if(connections, [&](const auto& connection) {
+          return connection.second.pipeline_executor.address() == msg.source;
+        });
+    TENZIR_ASSERT(connection != connections.end());
+    // FIXME: It is only safe to erase this connection _after_ there is no more
+    // in-flight async read for it. Otherwise, acquiring the mutex in it can
+    // throw an exception. The connections map likely has to point to shared
+    // pointers because of this.
+    connections.erase(connection);
+    if (not msg.reason) {
+      return;
+    }
+    // FIXME: Emit a diagnostic when a nested pipeline shuts down with an
+    // unexpected error.
+  }
+};
+
+template <class Elements>
+auto make_connection_manager(
+  typename connection_manager_actor<Elements>::template stateful_pointer<
+    connection_manager_state<Elements>>
+    self,
+  const load_tcp_args& args, const shared_diagnostic_handler& diagnostics,
+  bool is_hidden, const node_actor& node)
+  -> connection_manager_actor<Elements>::behavior_type {
+  self->state.self = self;
+  self->state.args = args;
+  self->state.diagnostics = diagnostics;
+  self->state.is_hidden = is_hidden;
+  self->state.node = node;
+  if (auto ok = self->state.start(); not ok) {
+    self->quit(std::move(ok.error()));
+    return connection_manager_actor<
+      Elements>::behavior_type::make_empty_behavior();
+  }
+  self->set_down_handler([self](const caf::down_msg& msg) {
+    self->state.handle_down_msg(msg);
+  });
+  return {
+    [self](atom::read, boost::asio::ip::tcp::socket::native_handle_type handle)
+      -> caf::result<chunk_ptr> {
+      return self->state.read_from_connection(handle);
+    },
+    [self](atom::write, Elements& elements) -> caf::result<void> {
+      return self->state.write_elements(std::move(elements));
+    },
+    [self](atom::read) -> caf::result<Elements> {
+      return self->state.read_elements();
+    },
+    [self](uint64_t op_index, uint64_t metric_index,
+           const type& schema) -> caf::result<void> {
+      (void)self;
+      // FIXME: Figure out how we can somehow forward custom metrics from the
+      // nested operator.
```

Comment:
I'm not sure if this is possible without major API changes. Need to discuss this.